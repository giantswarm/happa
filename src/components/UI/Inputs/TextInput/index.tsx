import {
  FormField,
  FormFieldProps,
  TextInput as Input,
  ThemeContext,
  ThemeType,
} from 'grommet';
import * as React from 'react';
import { css, useTheme } from 'styled-components';

const customTheme: ThemeType = {
  global: {
    drop: {
      background: 'input-background',
      shadowSize: 'none',
      border: {
        radius: '0px 0px 4px 4px',
      },
      zIndex: '9999',
      // @ts-expect-error
      extend: css`
        border: 1px solid ${({ theme }) => theme.global.colors.border.dark};
        font-size: ${({ theme }) => theme.text.small.size};
        font-weight: 400;
      `,
    },
  },
};

interface ITextInputProps extends React.ComponentPropsWithoutRef<typeof Input> {
  /**
   * The description text displayed above the input.
   */
  label?: FormFieldProps['label'];
  /**
   * Props to be passed to the text input wrapper.
   */
  contentProps?: FormFieldProps['contentProps'];
  /**
   * Props to be passed to the form field.
   */
  formFieldProps?: FormFieldProps & { className?: string };
  /**
   * Whether the input is required or not.
   */
  required?: FormFieldProps['required'];
  /**
   * An error to be displayed below the input.
   */
  error?: FormFieldProps['error'];
  /**
   * A description to be displayed below the input.
   */
  info?: FormFieldProps['info'];
  /**
   * A help text to be displayed between the label
   * and the input.
   */
  help?: FormFieldProps['help'];
  /**
   * The margin between the form field and other elements.
   */
  margin?: FormFieldProps['margin'];
  /**
   * Whether the input should have some extra padding or not.
   */
  pad?: FormFieldProps['pad'];
}

const TextInput = React.forwardRef<HTMLInputElement, ITextInputProps>(
  (
    {
      id,
      label,
      contentProps,
      formFieldProps,
      disabled,
      required,
      error,
      info,
      help,
      name,
      margin,
      pad,
      children,
      ...props
    },
    ref
  ) => {
    const theme = useTheme();
    const patchedContentProps = Object.assign({}, contentProps);
    if (!disabled) {
      patchedContentProps.background = theme.formField.background.color;
    }

    return (
      <ThemeContext.Extend value={customTheme}>
        <FormField
          htmlFor={id}
          label={label}
          contentProps={patchedContentProps}
          disabled={disabled}
          required={required}
          name={name}
          error={error}
          info={info}
          help={help}
          margin={margin}
          pad={pad}
          {...formFieldProps}
        >
          <Input
            {...props}
            id={id}
            ref={ref}
            disabled={disabled}
            required={required}
            name={name}
          />
          {children}
        </FormField>
      </ThemeContext.Extend>
    );
  }
);

TextInput.defaultProps = {
  size: 'medium',
  dropHeight: 'small',
};

export default TextInput;
