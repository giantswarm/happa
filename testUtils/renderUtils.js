import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
