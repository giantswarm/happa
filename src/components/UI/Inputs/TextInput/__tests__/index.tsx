import { renderWithTheme } from 'test/renderUtils';

import TextInput from '..';

describe('TextInput', () => {
  it('renders a simple input', () => {
    const { container } = renderWithTheme(TextInput, {
      value: 'Hi people',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a label', () => {
    const { container } = renderWithTheme(TextInput, {
      value: 'Hi people',
      label: 'Some input',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a validation error', () => {
    const { container } = renderWithTheme(TextInput, {
      value: 'Hi people',
      label: 'Some input',
      error: 'You failed at this',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a help message', () => {
    const { container } = renderWithTheme(TextInput, {
      value: 'Hi people',
      label: 'Some input',
      help: 'This is a very cool input',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with an info message', () => {
    const { container } = renderWithTheme(TextInput, {
      value: 'Hi people',
      label: 'Some input',
      info: 'You can set a cool value',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a disabled input', () => {
    const { container } = renderWithTheme(TextInput, {
      value: 'Hi people',
      label: 'Some input',
      disabled: true,
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
