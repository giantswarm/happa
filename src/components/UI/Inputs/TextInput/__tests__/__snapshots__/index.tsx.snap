// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders a disabled input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 yBlmn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-1qc9t9l-0 breHlv"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jazqMc FormField__FormFieldBox-m9hood-0 jGEMXj"
    >
      <label
        class="StyledText-sc-1sadyjn-0 esuSAl"
      >
        Some input
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 cFXDWg FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 dpaJQp"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 Obkjv"
            disabled=""
            value="Hi people"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TextInput renders a simple input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 yBlmn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-1qc9t9l-0 breHlv"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jazqMc FormField__FormFieldBox-m9hood-0 jGEMXj"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 jhJIyr FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 dpaJQp"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 jNnZQe"
            value="Hi people"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TextInput renders an input with a help message 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 yBlmn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-1qc9t9l-0 breHlv"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jazqMc FormField__FormFieldBox-m9hood-0 jGEMXj"
    >
      <label
        class="StyledText-sc-1sadyjn-0 esuSAl"
      >
        Some input
      </label>
      <span
        class="StyledText-sc-1sadyjn-0 erMGdB"
      >
        This is a very cool input
      </span>
      <div
        class="StyledBox-sc-13pk1d4-0 jhJIyr FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 dpaJQp"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 jNnZQe"
            value="Hi people"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TextInput renders an input with a label 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 yBlmn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-1qc9t9l-0 breHlv"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jazqMc FormField__FormFieldBox-m9hood-0 jGEMXj"
    >
      <label
        class="StyledText-sc-1sadyjn-0 esuSAl"
      >
        Some input
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 jhJIyr FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 dpaJQp"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 jNnZQe"
            value="Hi people"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TextInput renders an input with a validation error 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 yBlmn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-1qc9t9l-0 breHlv"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jazqMc FormField__FormFieldBox-m9hood-0 jGEMXj"
    >
      <label
        class="StyledText-sc-1sadyjn-0 esuSAl"
      >
        Some input
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 dfycIj FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 dpaJQp"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 jNnZQe"
            value="Hi people"
          />
        </div>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 dxYrxD"
      >
        You failed at this
      </span>
    </div>
  </div>
</div>
`;

exports[`TextInput renders an input with an info message 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 yBlmn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-1qc9t9l-0 breHlv"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jazqMc FormField__FormFieldBox-m9hood-0 jGEMXj"
    >
      <label
        class="StyledText-sc-1sadyjn-0 esuSAl"
      >
        Some input
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 jhJIyr FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 dpaJQp"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 jNnZQe"
            value="Hi people"
          />
        </div>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 fmFxNm"
      >
        You can set a cool value
      </span>
    </div>
  </div>
</div>
`;
